File: leetcode/lc-us/py/1626.py
from typing import List


class Solution:
    def bestTeamScore(self, scores: List[int], ages: List[int]) -> int:
        n = len(scores)
        v = sorted(zip(scores, ages), key=lambda x: (x[1], x[0]))
        f = [0] * n
        for i in range(n):
            f[i] = v[i][0]
            for j in range(0, i):
                if v[j][0] <= v[i][0] and v[j][1] < v[i][1] or v[j][1] == v[i][1]:
                    f[i] = max(f[i], f[j] + v[i][0])
        return max(f)

File: leetcode/lc-us/cpp/1626.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
    int tr[1000001]{}, max_age;

    int low_bit(int x) { return x & -x; }

    int query(int x) {
        int res = 0;
        while (x) res = max(res, tr[x]), x -= low_bit(x);
        return res;
    }

    void update(int x, int val) {
        while (x <= max_age) tr[x] = max(tr[x], val), x += low_bit(x);
    }

public:
    int bestTeamScore(vector<int> &scores, vector<int> &ages) {
        int n = (int) scores.size();
        max_age = *max_element(ages.begin(), ages.end());
        vector<pair<int, int>> v(n);
        for (int i = 0; i < n; i++) {
            v[i] = {scores[i], ages[i]};
        }
        sort(v.begin(), v.end());
        int res = 0;
        for (int i = 0; i < n; i++) {
            int score = v[i].first, age = v[i].second, cur = score + query(age);
            update(age, cur);
            res = max(res, cur);
        }
        return res;
    }
};